microprofile/cdi/src/main/java/io/helidon/microprofile/cdi/ContainerInstanceHolder.java

```diff
@@ -40,35 +44,52 @@ static void set(HelidonContainer container) {
     }
 
     // return true if the container was reset, indicating somebody started CDI by hand and then shut it down
-    static synchronized boolean isReset() {
-        return isReset;
+    static boolean isReset() {
+        return access(() -> isReset);
     }
 
-    static synchronized HelidonContainer get() {
-        HelidonContainer helidonContainer = CONTAINER.get();
-        if (null == helidonContainer) {
-            helidonContainer = fromBuildTimeInitializer();
-            CONTAINER.compareAndSet(null, helidonContainer);
-        }
-        return helidonContainer;
+    static HelidonContainer get() {
+        return access(() -> {
+            HelidonContainer helidonContainer = CONTAINER.get();
+            if (null == helidonContainer) {
+                helidonContainer = fromBuildTimeInitializer();
+                CONTAINER.compareAndSet(null, helidonContainer);
+            }
+            return helidonContainer;
+        });
     }
 
     private static HelidonContainer fromBuildTimeInitializer() {
         // this is the time the class should get initialized
         return BuildTimeInitializer.get();
     }
 
-    static synchronized void addListener(Runnable runnable) {
-        RESET_LISTENERS.add(runnable);
+    static void addListener(Runnable runnable) {
+        access(() -> {
+            RESET_LISTENERS.add(runnable);
+            return null;
+        });
+    }
+
+    static void reset() {
+        access(() -> {
+            isReset = true;
+            CONTAINER.set(null);
+            for (Runnable resetListener : RESET_LISTENERS) {
+                resetListener.run();
+            }
+            HelidonCdiProvider.unset();
+            RESET_LISTENERS.clear();
+            return null;
+        });
     }
 
-    static synchronized void reset() {
-        isReset = true;
-        CONTAINER.set(null);
-        for (Runnable resetListener : RESET_LISTENERS) {
-            resetListener.run();
+    private static <T> T access(Supplier<T> operation) {
+        try {
+            ACCESS_GUARD.writeLock().lock();
```

Comment:
As noted above, since I never saw a `readLock()` operation, you probably only need a `ReentrantLock`.

Reply:
done